File: src/AltV.Net/Elements/Entities/Player.cs
using System;
using System.Runtime.InteropServices;
using AltV.Net.Data;
using AltV.Net.Elements.Args;
using AltV.Net.Native;

namespace AltV.Net.Elements.Entities
{
    public class Player : Entity, IPlayer
    {
        public static ushort GetId(IntPtr playerPointer) => AltVNative.Player.Player_GetID(playerPointer);

        public override uint Model
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetModel(NativePointer);
            }
        }

        public override Position Position
        {
            get
            {
                CheckExistence();
                var position = Position.Zero;
                AltVNative.Player.Player_GetPosition(NativePointer, ref position);
                return position;
            }
            set
            {
                CheckExistence();
                AltVNative.Player.Player_SetPosition(NativePointer, value);
            }
        }

        public override Rotation Rotation
        {
            get
            {
                CheckExistence();
                var rotation = Rotation.Zero;
                AltVNative.Player.Player_GetRotation(NativePointer, ref rotation);
                return rotation;
            }
            set
            {
                CheckExistence();
                AltVNative.Player.Player_SetRotation(NativePointer, value);
            }
        }

        public override short Dimension
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetDimension(NativePointer);
            }
            set
            {
                CheckExistence();
                AltVNative.Player.Player_SetDimension(NativePointer, value);
            }
        }

        public uint Ping
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetPing(NativePointer);
            }
        }

        public override void SetMetaData(string key, object value)
        {
            CheckExistence();
            var mValue = MValue.CreateFromObject(value);
            AltVNative.Player.Player_SetMetaData(NativePointer, key, ref mValue);
        }

        public override bool GetMetaData<T>(string key, out T result)
        {
            CheckExistence();
            var mValue = MValue.Nil;
            AltVNative.Player.Player_GetMetaData(NativePointer, key, ref mValue);
            if (!(mValue.ToObject() is T cast))
            {
                result = default;
                return false;
            }

            result = cast;
            return true;
        }

        public override void SetSyncedMetaData(string key, object value)
        {
            CheckExistence();
            var mValue = MValue.CreateFromObject(value);
            AltVNative.Player.Player_SetSyncedMetaData(NativePointer, key, ref mValue);
        }

        public override bool GetSyncedMetaData<T>(string key, out T result)
        {
            CheckExistence();
            var mValue = MValue.Nil;
            AltVNative.Player.Player_GetSyncedMetaData(NativePointer, key, ref mValue);
            if (!(mValue.ToObject() is T cast))
            {
                result = default;
                return false;
            }

            result = cast;
            return true;
        }

        public bool IsConnected
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsConnected(NativePointer);
            }
        }

        public string Name
        {
            get
            {
                CheckExistence();
                var ptr = IntPtr.Zero;
                AltVNative.Player.Player_GetName(NativePointer, ref ptr);
                return Marshal.PtrToStringAnsi(ptr);
            }
            set
            {
                CheckExistence();
                AltVNative.Player.Player_SetName(NativePointer, value);
            }
        }

        public ushort Health
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetHealth(NativePointer);
            }
            set
            {
                CheckExistence();
                AltVNative.Player.Player_SetHealth(NativePointer, value);
            }
        }

        public bool IsDead
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsDead(NativePointer);
            }
        }

        public bool IsJumping
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsJumping(NativePointer);
            }
        }

        public bool IsInRagdoll
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsInRagdoll(NativePointer);
            }
        }

        public bool IsAiming
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsAiming(NativePointer);
            }
        }

        public bool IsShooting
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsShooting(NativePointer);
            }
        }

        public bool IsReloading
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsReloading(NativePointer);
            }
        }

        public ushort Armor
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetArmor(NativePointer);
            }
            set
            {
                CheckExistence();
                AltVNative.Player.Player_SetArmor(NativePointer, value);
            }
        }

        public float MoveSpeed
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetMoveSpeed(NativePointer);
            }
        }

        public uint Weapon
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetWeapon(NativePointer);
            }
        }

        public ushort Ammo
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetAmmo(NativePointer);
            }
        }

        public Position AimPosition
        {
            get
            {
                CheckExistence();
                var position = Position.Zero;
                AltVNative.Player.Player_GetAimPos(NativePointer, ref position);
                return position;
            }
        }

        public Rotation HeadRotation
        {
            get
            {
                CheckExistence();
                var rotation = Rotation.Zero;
                AltVNative.Player.Player_GetHeadRotation(NativePointer, ref rotation);
                return rotation;
            }
        }

        public bool IsInVehicle
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_IsInVehicle(NativePointer);
            }
        }

        public IVehicle Vehicle
        {
            get
            {
                CheckExistence();
                var entityPointer = AltVNative.Player.Player_GetVehicle(NativePointer);
                if (entityPointer == IntPtr.Zero) return null;
                return Alt.Module.VehiclePool.GetOrCreate(entityPointer, out var vehicle) ? vehicle : null;
            }
        }

        public byte Seat
        {
            get
            {
                CheckExistence();
                return AltVNative.Player.Player_GetSeat(NativePointer);
            }
        }

        public Player(IntPtr nativePointer, ushort id) : base(nativePointer, BaseObjectType.Player, id)
        {
        }

        public void Spawn(Position position)
        {
            CheckExistence();
            AltVNative.Player.Player_Spawn(NativePointer, position);
        }

        public void Despawn()
        {
            CheckExistence();
            AltVNative.Player.Player_Despawn(NativePointer);
        }

        public void SetDateTime(int day, int month, int year, int hour, int minute, int second)
        {
            CheckExistence();
            AltVNative.Player.Player_SetDateTime(NativePointer, day, month, year, hour, minute, second);
        }

        public void SetWeather(uint weather)
        {
            CheckExistence();
            AltVNative.Player.Player_SetWeather(NativePointer, weather);
        }

        public void Kick(string reason)
        {
            CheckExistence();
            AltVNative.Player.Player_Kick(NativePointer, reason);
        }

        public void Emit(string eventName, params object[] args)
        {
            CheckExistence();
            Alt.Server.TriggerClientEvent(this, eventName, args);
        }

        public ReadOnlyPlayer Copy()
        {
            CheckExistence();
            var readOnlyPlayer = ReadOnlyPlayer.Empty;
            AltVNative.Player.Player_Copy(NativePointer, ref readOnlyPlayer);
            return readOnlyPlayer;
        }
    }
}

File: clrhost/src/altv-c-api/player.cpp
#include "player.h"

// Entity

uint16_t Player_GetID(alt::IPlayer* player) {
    return player->GetID();
}

uint32_t Player_GetModel(alt::IPlayer* player) {
    return player->GetModel();
}

void Player_GetPosition(alt::IPlayer* player, position_t& position) {
    auto playerPosition = player->GetPosition();
    position.x = playerPosition.x;
    position.y = playerPosition.y;
    position.z = playerPosition.z;
}

void Player_SetPosition(alt::IPlayer* player, alt::Position pos) {
    player->SetPosition(pos);
}

void Player_GetRotation(alt::IPlayer* player, rotation_t& rotation) {
    auto playerRotation = player->GetRotation();
    rotation.roll = playerRotation.roll;
    rotation.pitch = playerRotation.pitch;
    rotation.yaw = playerRotation.yaw;
}

void Player_SetRotation(alt::IPlayer* player, alt::Rotation rot) {
    player->SetRotation(rot);
}

int16_t Player_GetDimension(alt::IPlayer* player) {
    return player->GetDimension();
}

void Player_SetDimension(alt::IPlayer* player, uint16_t dimension) {
    player->SetDimension(dimension);
}

void Player_GetMetaData(alt::IPlayer* player, const char* key, alt::MValue &val) {
    val = player->GetMetaData(key);
}

void Player_SetMetaData(alt::IPlayer* player, const char* key, alt::MValue* val) {
    player->SetMetaData(key, *val);
}

void Player_GetSyncedMetaData(alt::IPlayer* player, const char* key, alt::MValue &val) {
    val = player->GetSyncedMetaData(key);
}

void Player_SetSyncedMetaData(alt::IPlayer* player, const char* key, alt::MValue* val) {
    player->SetSyncedMetaData(key, *val);
}

// Player

bool Player_IsConnected(alt::IPlayer* player) {
    return player->IsConnected();
}

void Player_Spawn(alt::IPlayer* player, alt::Position pos) {
    player->Spawn(pos);
}

void Player_Despawn(alt::IPlayer* player) {
    player->Despawn();
}

void Player_GetName(alt::IPlayer* player, const char*&name) {
    name = player->GetName().CStr();
}

void Player_SetName(alt::IPlayer* player, const char* name) {
    player->SetName(name);
}

uint16_t Player_GetHealth(alt::IPlayer* player) {
    return player->GetHealth();
}

void Player_SetHealth(alt::IPlayer* player, uint16_t health) {
    player->SetHealth(health);
}

void Player_SetDateTime(alt::IPlayer* player, int day, int month, int year, int hour, int minute, int second) {
    player->SetDateTime(day, month, year, hour, minute, second);
}

void Player_SetWeather(alt::IPlayer* player, uint32_t weather) {
    player->SetWeather(weather);
}

bool Player_IsDead(alt::IPlayer* player) {
    return player->IsDead();
}

bool Player_IsJumping(alt::IPlayer* player) {
    return player->IsJumping();
}

bool Player_IsInRagdoll(alt::IPlayer* player) {
    return player->IsInRagdoll();
}

bool Player_IsAiming(alt::IPlayer* player) {
    return player->IsAiming();
}

bool Player_IsShooting(alt::IPlayer* player) {
    return player->IsShooting();
}

bool Player_IsReloading(alt::IPlayer* player) {
    return player->IsReloading();
}

uint16_t Player_GetArmor(alt::IPlayer* player) {
    return player->GetArmor();
}

void Player_SetArmor(alt::IPlayer* player, uint16_t armor) {
    player->SetArmor(armor);
}

float Player_GetMoveSpeed(alt::IPlayer* player) {
    return player->GetMoveSpeed();
}

uint32_t Player_GetWeapon(alt::IPlayer* player) {
    return player->GetWeapon();
}

uint16_t Player_GetAmmo(alt::IPlayer* player) {
    return player->GetAmmo();
}

void Player_GetAimPos(alt::IPlayer* player, position_t& aimPosition) {
    auto playerAimPosition = player->GetAimPos();
    aimPosition.x = playerAimPosition.x;
    aimPosition.y = playerAimPosition.y;
    aimPosition.z = playerAimPosition.z;
}

void Player_GetHeadRotation(alt::IPlayer* player, rotation_t& headRotation) {
    auto playerHeadRotation = player->GetHeadRotation();
    headRotation.roll = playerHeadRotation.roll;
    headRotation.pitch = playerHeadRotation.pitch;
    headRotation.yaw = playerHeadRotation.yaw;
}

bool Player_IsInVehicle(alt::IPlayer* player) {
    return player->IsInVehicle();
}

alt::IVehicle* Player_GetVehicle(alt::IPlayer* player) {
    return player->GetVehicle();
}

uint8_t Player_GetSeat(alt::IPlayer* player) {
    return player->GetSeat();
}

void Player_Kick(alt::IPlayer* player, const char* reason) {
    player->Kick(reason);
}

uint32_t Player_GetPing(alt::IPlayer* player) {
    return player->GetPing();
}

void Player_Copy(alt::IPlayer* player, player_struct_t* player_struct) {
    player_struct->id = player->GetID();
    player_struct->type = player->GetType();
    auto position = player->GetPosition();
    player_struct->position.x = position.x;
    player_struct->position.y = position.y;
    player_struct->position.z = position.x;
    auto rotation = player->GetRotation();
    player_struct->rotation.roll = rotation.roll;
    player_struct->rotation.pitch = rotation.pitch;
    player_struct->rotation.yaw = rotation.yaw;
    player_struct->dimension = player->GetDimension();
}